File: Javascript/src/01.AlgorithmicToolbox/Week4DivideAndConquer/closestPoints.js
/**
 * Problem 6 - Closest Points
 * Given n points on a plane, find the smallest distance between a pair
 * of two different points.
 * @param {Array.<{x: number, y: number}>} points set of points on a plane
 * @returns {number} smallest distance
 */
export function closestPointsDistance(points) {
  // sort according to x-coordinate
  points.sort((pointA, pointB) => pointA.x - pointB.x);
  return closestPointsDistanceHelper(points, 0, points.length);
}

function closestPointsDistanceHelper(points, startIndex, endIndex) {
  if (endIndex - startIndex === 1) {
    return Infinity;
  }
  // find closest distance of each half
  const middleIndex = startIndex + Math.floor((endIndex - startIndex) / 2);
  const leftClosestPointsDistance = closestPointsDistanceHelper(
    points,
    startIndex,
    middleIndex
  );
  const rightClosestPointsDistance = closestPointsDistanceHelper(
    points,
    middleIndex,
    endIndex
  );
  let closestPointsDistance = Math.min(
    leftClosestPointsDistance,
    rightClosestPointsDistance
  );
  // collect points in middle strip within closestPointsDistance to middle point
  const middlePoint = points[middleIndex];
  const pointsInMiddleStrip = [];
  for (let i = startIndex; i < endIndex; ++i) {
    if (Math.abs(points[i].x - middlePoint.x) < closestPointsDistance) {
      pointsInMiddleStrip.push(points[i]);
    }
  }
  // sort middle points by y coordinates
  pointsInMiddleStrip.sort((pointA, pointB) => pointA.y - pointB.y);
  // compare each point with 7 of its neighbors, updating closest points distance
  for (let i = 0; i < pointsInMiddleStrip.length; ++i) {
    for (let j = i + 1; j < i + 8 && j < pointsInMiddleStrip.length; ++j) {
      const distance = distanceBetweenPoints(
        pointsInMiddleStrip[i],
        pointsInMiddleStrip[j]
      );
      if (distance < closestPointsDistance) {
        closestPointsDistance = distance;
      }
    }
  }
  return closestPointsDistance;
}

function distanceBetweenPoints(pointA, pointB) {
  return Math.sqrt(
    Math.pow(pointA.x - pointB.x, 2) + Math.pow(pointA.y - pointB.y, 2)
  );
}

File: Rust/src/_01_algorithmic_toolbox/week4_divide_and_conquer/closest_points.rs
///
/// Problem 6 - Closest Points
/// Given n points on a plane, find the smallest distance between a pair
/// of two different points.
pub fn closest_points(points: &mut [Point]) -> f64 {
    points.sort_by(|point_a, point_b| point_a.x.partial_cmp(&point_b.x).unwrap());
    closest_points_helper(points)
}
fn closest_points_helper(points: &[Point]) -> f64 {
    if points.len() == 1 {
        return f64::INFINITY;
    }
    // find closest distance of each half
    let middle_index = points.len() / 2;
    let left_closest_points_distance = closest_points_helper(&points[..middle_index]);
    let right_closest_points_distance = closest_points_helper(&points[middle_index..]);
    let mut closest_points_distance =
        left_closest_points_distance.min(right_closest_points_distance);
    // collect points in middle strip within closest points distance to middle point
    let middle_point = &points[middle_index];
    let mut points_in_middle_strip = vec![];
    for point in points {
        if (point.x - middle_point.x).abs() < closest_points_distance {
            points_in_middle_strip.push(point)
        }
    }
    // sort middle points by y coordinates
    points_in_middle_strip.sort_by(|point_a, point_b| point_a.y.partial_cmp(&point_b.y).unwrap());
    // compare each point with 7 of its neighbors, updating closest points distance
    for i in 0..points_in_middle_strip.len() {
        for j in i + 1..(i + 8).min(points_in_middle_strip.len()) {
            if j > points_in_middle_strip.len() {
                break;
            }
            let distance =
                distance_between_points(points_in_middle_strip[i], points_in_middle_strip[j]);
            closest_points_distance = closest_points_distance.min(distance);
        }
    }
    closest_points_distance
}

pub struct Point {
    x: f64,
    y: f64,
}

fn distance_between_points(point_a: &Point, point_b: &Point) -> f64 {
    ((point_a.x - point_b.x).powi(2) + (point_a.y - point_b.y).powi(2)).sqrt()
}

#[cfg(test)]
mod test {
    use super::*;

    fn approx_equal(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }
    #[test]
    fn returns_closest_points_distance() {
        assert_eq!(
            approx_equal(
                closest_points(&mut vec![
                    Point { x: 0.0, y: 0.0 },
                    Point { x: 3.0, y: 4.0 }
                ]),
                5.0
            ),
            true
        );
        assert_eq!(
            approx_equal(
                closest_points(&mut vec![
                    Point { x: 7.0, y: 7.0 },
                    Point { x: 1.0, y: 100.0 },
                    Point { x: 4.0, y: 8.0 },
                    Point { x: 7.0, y: 7.0 }
                ]),
                0.0
            ),
            true
        );
        assert_eq!(
            approx_equal(
                closest_points(&mut vec![
                    Point { x: 4.0, y: 4.0 },
                    Point { x: -2.0, y: -2.0 },
                    Point { x: -3.0, y: -4.0 },
                    Point { x: -1.0, y: 3.0 },
                    Point { x: 2.0, y: 3.0 },
                    Point { x: -4.0, y: 0.0 },
                    Point { x: 1.0, y: 1.0 },
                    Point { x: -1.0, y: -1.0 },
                    Point { x: 3.0, y: -1.0 },
                    Point { x: -4.0, y: 2.0 },
                    Point { x: -2.0, y: 4.0 },
                ]),
                1.414213
            ),
            true
        );
    }
}
